motd/src/lib.rs

```diff
@@ -43,3 +255,454 @@ impl std::fmt::Display for Error {
+    //         resp array to be a pointer to an array of PamResponses, rather than
+    //         an array of PamResponse pointers. This means the double indirection
+    //         is in order for us to write to the output variable, so the caller
+    //         is responsible for making sure there is a word there for us to write
+    //         to.
+    unsafe {
+        *resp = {
+            let alloc_size = mem::size_of::<pam_sys::PamResponse>() * num_msg as usize;
+            let resp_buf = libc::malloc(alloc_size);
+            slice::from_raw_parts_mut(resp_buf as *mut u8, alloc_size).fill(0);
+            mem::transmute(resp_buf)
+        };
+    }
+
+    // Safety: we cast from a `&ConvData` to a `*mut c_void` when creating the PamConversation,
+    //         so we are casting to the right type. The assertion against null means that
+    //         it is safe to convert to a reference.
+    assert!(!appdata_ptr.is_null());
+    let conv_data = unsafe { &mut *mem::transmute::<_, *mut ConvData>(appdata_ptr) };
+
+    for i in 0..(num_msg as usize) {
+        // Safety: the caller is responsible for giving us a complete message list.
+        //         Any issue with this operation would be due to an issue with how the
+        //         caller has set things up since the loop means we are in-bounds.
+        let msg = unsafe { *msgs[i] };
+
+        let msg_style = pam_sys::types::PamMessageStyle::from(msg.msg_style);
+        match msg_style {
+            PamMessageStyle::PROMPT_ECHO_OFF => {
+                conv_data
+                    .errs
+                    .push(merr!("pam_motd.so asked for a password"));
+            }
+            PamMessageStyle::PROMPT_ECHO_ON => {
+                conv_data
+                    .errs
+                    .push(merr!("pam_motd.so asked for a username"));
+            }
+            PamMessageStyle::TEXT_INFO => {
+                // Safety: pam_motd.so will give us a valid cstring here.
+                let msg = unsafe { ffi::CStr::from_ptr(msg.msg) };
+                match msg.to_str() {
+                    Ok(s) => conv_data.motd_msg.push_str(s),
+                    Err(e) => conv_data
+                        .errs
+                        .push(merr!("err converting motd chunk: {}", e)),
+                }
+            }
+            PamMessageStyle::ERROR_MSG => {
+                // Safety: pam_motd.so will give us a valid cstring here.
+                let msg = unsafe { ffi::CStr::from_ptr(msg.msg) };
+                match msg.to_str() {
+                    Ok(s) => conv_data.errs.push(merr!("pam_mod.so says '{}'", s)),
+                    Err(e) => conv_data
+                        .errs
+                        .push(merr!("err converting motd err msg: {}", e)),
+                }
+            }
+        }
+    }
+
+    PamReturnCode::SUCCESS as libc::c_int
+}
+
+//
+// .so discovery logic
+//
+// TODO(ethan): I should try to resolve the location of pam_motd.so
```

Comment:
Annoyingly, this ends up being distro-specific.

Reply:
That's super annoying. Maybe it is possible to make a .so file which depends on `pam_motd.so`, then write it out to a tmp file and parse the output of `ldd` when run on it. That should be somewhat distro agnostic right?

Reply:
Maybe, but pam_motd.so doesn't show up in the output of `ldconfig -p`, so maybe that won't work at all. I was hoping we could just grep for pam_motd.so in that output, but... pam_motd.so is in a security subdir and subdirs don't seem to be scanned by ldconfig, only files directly in the configuration of /etc/ld.so.conf and includes in /etc/ld.so.conf.d.

Reply:
Dang, that's super annoying. It makes sense though since the pam modules are not really regular shared libraries. It's just that the .so format makes for a good way to bundle up arbitrary code to load at runtime.